### PR TITLE
Refactor hints

### PR DESCRIPTION
**Pull request by fmoko**
Breaking change. This removes hints from the end of files, and puts them into `info.toml`. You can now access hints using:

```
rustlings hint <exerciseName>
```

ALSO this changes the exercise system to index by name for `run` and `hint`, so:

```
rustlings run exercises/if/if1.rs
```

becomes

```
rustlings run if1
```

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -55,26 +60,38 @@ fn main() {
     let exercises = toml::from_str::<ExerciseList>(toml_str).unwrap().exercises;
 
     if let Some(ref matches) = matches.subcommand_matches("run") {
-        let filename = matches.value_of("file").unwrap_or_else(|| {
-            println!("Please supply a file name!");
+        let name = matches.value_of("name").unwrap_or_else(|| {
```
Comment (jrvidal) [inline src/main.rs]: Doesn't `clap` already validate that this is present?

File: src/main.rs [hunk 2]
```diff
@@ -55,26 +60,38 @@ fn main() {
...
-            Path::new(filename)
-                .canonicalize()
-                .map(|p| p.ends_with(&e.path))
-                .unwrap_or(false)
-        };
+        let matching_exercise = |e: &&Exercise| name == e.name;
 
         let exercise = exercises.iter().find(matching_exercise).unwrap_or_else(|| {
-            println!("No exercise found for your file name!");
+            println!("No exercise found for your given name!");
```
Comment (jrvidal) [inline src/main.rs]: Now that we rely on exercise names, I wonder whether there should be a test to check that exercise names are unique.

File: exercises/error_handling/errorsn.rs [hunk 3]
```diff
@@ -13,7 +13,7 @@
 // type goes where the question marks are, and how do we return
 // that type from the body of read_and_validate?
 //
-// Scroll down for hints :)
+// Execute `rustlings hint errors4` for hints :)
```
Comment (jrvidal) [inline exercises/error_handling/errorsn.rs]: `errorsn`

**Comment by fmoko** [comment]
Also closes #203.

File: exercises/error_handling/result1.rs [hunk 4]
```diff
@@ -1,5 +1,5 @@
 // result1.rs
-// Make this test pass! Scroll down for hints :)
+// Make this test pass! Execute `rustlings hint option2` for hints :)
```
Comment (jrvidal) [inline exercises/error_handling/result1.rs]: `result1`

File: exercises/functions/functions1.rs [hunk 5]
```diff
@@ -1,46 +1,8 @@
 // functions1.rs
-// Make me compile! Scroll down for hints :)
+// Make me compile! Execute `rustlings hint function1` for hints :)
```
Comment (jrvidal) [inline exercises/functions/functions1.rs]: Typo in function-_s_

File: info.toml [hunk 6]
```diff
@@ -1,241 +1,612 @@
...
 
 [[exercises]]
+name = "errors3"
 path = "exercises/error_handling/errors3.rs"
 mode = "test"
+hint = """
+If other functions can return a `Result`, why shouldn't `main`?"""
 
 [[exercises]]
+name = "errors4"
```
Comment (jrvidal) [inline info.toml]: `errorsn`?

File: info.toml [hunk 7]
```diff
@@ -1,241 +1,612 @@
...
 mode = "test"
+hint = """
+Try using a `match` statement where the arms are `Some(thing)` and `None`.
+Or set a default value to print out if you get `None` by using the
+function `unwrap_or`.
+Or use an `if let` statement on the result of `pop()` to both destructure
+a `Some` value and only print out something if we have a value!"""
 
 [[exercises]]
+name = "option2"
```
Comment (jrvidal) [inline info.toml]: `result1`?

File: src/main.rs (hunk 2 above)
Comment (fmoko) [reply src/main.rs]: Probably, but it'd be out of scope for this PR

**Pull request merged**